node/network/bridge/src/multiplexer.rs

```diff
@@ -0,0 +1,133 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::pin::Pin;
+
+use futures::channel::mpsc;
+use futures::stream::Stream;
+use futures::task::{Context, Poll};
+use strum::IntoEnumIterator;
+
+use parity_scale_codec::{Decode, Error as DecodingError};
+
+use sc_network::config as network;
+use sc_network::PeerId;
+
+use polkadot_node_network_protocol::request_response::{
+	request::IncomingRequest, v1, Protocol, RequestResponseConfig,
+};
+use polkadot_subsystem::messages::AllMessages;
+
+/// Multiplex incoming network requests.
+///
+/// This multiplexer consumes all request streams and makes them a `Stream` of a single message
+/// type, useful for the network bridge to send them via the `Overseer` to other subsystems.
+pub struct RequestMultiplexer {
+	receivers: Vec<(Protocol, mpsc::Receiver<network::IncomingRequest>)>,
+	next_poll: usize,
+}
+
+/// Multiplexing can fail in case of invalid messages.
+pub struct RequestMultiplexError {
+	/// The peer that sent the invalid message.
+	pub peer: PeerId,
+	/// The error that occurred.
+	pub error: DecodingError,
+}
+
+impl RequestMultiplexer {
+	/// Create a new `RequestMultiplexer`.
+	///
+	/// This function uses `Protocol::get_config` for each available protocol and creates a
+	/// `RequestMultiplexer` from it. The returned `RequestResponseConfig`s must be passed to the
+	/// network implementation.
+	pub fn new() -> (Self, Vec<RequestResponseConfig>) {
+		let (receivers, cfgs): (Vec<_>, Vec<_>) = Protocol::iter()
+			.map(|p| {
+				let (rx, cfg) = p.get_config();
+				((p, rx), cfg)
+			})
+			.unzip();
+
+		(
+			Self {
+				receivers,
+				next_poll: 0,
+			},
+			cfgs,
+		)
+	}
+}
+
+impl Stream for RequestMultiplexer {
+	type Item = Result<AllMessages, RequestMultiplexError>;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+		let len = self.receivers.len();
+		let mut count = len;
+		let mut i = self.next_poll;
+		let mut result = Poll::Ready(None);
+		// Poll streams in round robin fashion:
+		while count > 0 {
+			// % safe, because count initialized to len, loop would not be entered if 0:
+			let (p, rx): &mut (_, _) = &mut self.receivers[i % len];
+			i += 1;
+			count -= 1;
+			match Pin::new(rx).poll_next(cx) {
+				// If at least one stream is pending, we are pending as well:
+				Poll::Pending => result = Poll::Pending,
+				// Receiver is a fused stream, which allows for this simple handling of 
+				// exhausted ones.
+				Poll::Ready(None) => {}
```

Comment:
Is that expected that one of receivers streams can end? 
If not, maybe we should abort in this case by returning `Poll::Ready(None)` instead of silently ignoring it?
Aso, is it safe to poll it if `is_terminated` is true? Not sure.

Reply:
Receiver is a fused Stream, so it should be safe. Not sure about the first question. It seemed logical that the stream ends if all its sources are exhausted.

The only reason those streams should get exhausted is because we are shutting down, in that case it might or might not matter to deliver any messages still in the queues. It felt a bit safer to try to deliver them, until everything is exhausted. On the other hand, receivers will hardly be able to respond anyways ....

I guess that's an edge case, which should not make much of a difference, either way.

Reply:
> Receiver is a fused Stream, so it should be safe.

it still can panic:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f1ac00093b376fc6cefb41033adff814

Reply:
oh boy, thanks a lot! I could have sworn that it is safe to poll a fused stream after exhaustion.

Reply:
I think the safest thing would be to early return on the first closed stream, i.e. return `Ready(None)`.

Reply:
Hold on, it only happens with `.next()`, `.fuse()` on a stream guarantees that `.poll_next()` will not panic. So depending of the use case here https://docs.rs/futures/0.3.8/futures/stream/trait.StreamExt.html#method.fuse

So what you want is: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=691bfc7175278e2887340f5289df63d3

Reply:
Ok, so I confused a Stream implementing `FusedStream` with a stream that got fused by means of `fuse`. Ok this definitely deserves a test. Thank you guys. Will provide a fix in a minute.